### PR TITLE
Fix overload derivative decl not being added to codegen

### PR DESCRIPTION
**Pull request by sudo-panda**
- Fixes linker not able to find overload code
- Fixes `setPreviousDecl` called for wrong decl

### CODE REVIEW

File: lib/Differentiator/DerivativeBuilder.cpp [hunk 1]
```diff
@@ -45,10 +45,14 @@ namespace clad {
     // otherwise it would end up in the LookupResult.
     derivedFD->getDeclContext()->addDecl(derivedFD);
 
-    if (R.empty())
-      return;
-    // Register the function on the redecl chain.
-    derivedFD->setPreviousDecl(cast<FunctionDecl>(R.getFoundDecl()));
+    for (NamedDecl* I : R) {
```
Comment (vgvassilev) [inline lib/Differentiator/DerivativeBuilder.cpp:48]: I think we should either try using `Sema::CheckFunctionDeclaration` and if that does not work we should use `ASTContext::hasSameFunctionTypeIgnoringExceptionSpec(FD->getType(), derivedFD->getType())`

Comment (sudo-panda) [reply lib/Differentiator/DerivativeBuilder.cpp:48]: Alright!

File: lib/Differentiator/VisitorBase.cpp [hunk 2]
```diff
@@ -573,31 +573,21 @@ namespace clad {
...
+        Create(m_Context, thisExpr, /*isArrow=*/true, noLoc, NNS, noLoc, FD,
+               DAP, FD->getNameInfo(),
+               /*TemplateArgs=*/nullptr, m_Context.BoundMemberTy,
+               ExprValueKind::VK_RValue,
+               ExprObjectKind::OK_Ordinary
+               CLAD_COMPAT_CLANG9_MemberExpr_ExtraParams(NOUR_None));
+    return m_Sema
+        .BuildCallToMemberFunction(getCurrentScope(), memberExpr, noLoc,
+                                   argExprs, noLoc)
+        .get();
```
Comment (sudo-panda) [inline lib/Differentiator/VisitorBase.cpp:590]: @parth-07 could you please see if this implementation fails in any cases that you need it for.

Comment (parth-07) [reply lib/Differentiator/VisitorBase.cpp:590]: No, this implementation should work fine for cases I needed this function for.

Comment (sudo-panda) [reply lib/Differentiator/VisitorBase.cpp:590]: Thanks!

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Comment by vgvassilev** [comment]
codecov seems confused again.

**Pull request merged**